L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,81 @@
+Title
+----
+* Author(s): James Newton-King
+* Approver: a11r
+* Status: Draft
+* Implemented in: csharp
+* Last updated: 2021-05-19
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+grpc-dotnet currently has no load balancing support. A channel is always created for a single address.
+
+This documentation outlines an initial load balancing implementation for the grpc-dotnet client.
+
+## Background
+
+grpc-dotnet currently doesn't support any form of load balancing in the client. The changes outlined in this proposal will enable basic gRPC load balancing and some of its related features. These include:
+
+* Name resolving. Built-in support is included for resolving from DNS, and a static resolver. Custom resolvers can be created.
+* Load balancing strategies. Built-in support is included for pick first and round robin load balancing. Custom load balancers can be created.
+* Connectivity status at sub-channel and channel level
+* Wait for ready
+
+## Proposal
+
+![Diagram](L80_graphics/diagram.png)
+
+### Initialization
+
+A channel is created with an address, an optional service provider, and an optional service config.
+
+```csharp
+var channel = GrpcChannel.ForAddress(
+    "dns://backend.default.svc.cluster.local",
+    new GrpcChannelOptions { ChannelCredentials = ChannelCredentials.Insecure });
+var client = new Greet.GreeterClient(channel);
+
+var response = await client.SayHelloAsync(new HelloRequest { Name = "world" });
+Console.WriteLine(response.Message);
+```
+
+* Address scheme will be used to create a resolver.
+  * Existing logic allows for `http` and `https`. They are special-cased to never use a resolver.
+  * A collection of resolver factories are fetched from the service provider. One is selected using the address scheme and it will create resolver. For example, `dns` will match to `DnsAddressResolverFactory` which will create a `DnsAddressResolver`.
+* Load balancer name in service configure will be used to create a load balancer.
+  * A collection of load balancer factories are fetched from the service provider. One is selected using the configured name (if none specified then `pick_first` is used) and it will create load balancer.
+  * Load balancer is initialized with an `IChannelController` instance by the channel. This is used to call back to the channel to create sub-channels and update the picker.
+
+### Connecting
+
+* No network activity until `GrpcChannel.ConnectAsync` is called, or the first RPC is started.
+* On connect, the `ConnectionManager` will subscribe to address resolver and call the resolver's `RefreshAsync` to prompt the resolver to run immediately.
```

Comment:
Resolver's have `RefreshAsync` method to ask them to refresh immediately. For example, DNS resolver will refresh on a timer every 5 seconds, and `RefreshAsync` will also cause the resolver to immediately get updated DNS results.

There is the `IChannelController.ResolveNowAsync` method. It sounds like it is the same as `RequestReresolution`. The load balancer policy can use it to resolve now if required.

Reply:
I believe the current subchannel connect logic in .NET will request an update from the resolver if there is more than one error when connecting to a subchannel.

That lines up with what you said here: https://github.com/grpc/grpc/issues/12295#issuecomment-375098264

If that is the case then failing subchannels will automatically update the resolved addresses. The situation where updating on an interval is useful would only be when a server has scaled up and there are more addresses that the client isn't aware of.

I'll make polling an option, with the default as off.

Reply:
> I believe the current subchannel connect logic in .NET will request an update from the resolver if there is more than one error when connecting to a subchannel.

This decision should be made by the LB policy, not by the channel or subchannel.  There are some cases where an LB policy may not want to re-resolve just because an individual subchannel becomes disconnected.

(I believe that Java and Go may still have some legacy behavior where the channel is making this decision, but our long-term intent is to eliminate such behavior.)

> I'll make polling an option, with the default as off.

Sounds good.  CC @ejona86 @dfawley as an FYI.

Reply:
I've updated it so that load balancers request re-resolve, not the sub-channel.